### PR TITLE
Add module for handling version information of stdlib

### LINKED ISSUE #576: stdlib version information function
### Motivation

It would be good to have a function `stdlib_version()`

```
function stdlib_version() result(r)
    character(len=:), allocatable :: r
    r = "v0.1.0"
end function
```

This information is useful for external build tools, and for bug reporting.

### Prior Art

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by awvwgk**
Closes #576

### CODE REVIEW

File: doc/specs/stdlib_version.md
Comment (ivan-pi) [inline doc/specs/stdlib_version.md:12]:
```suggestion
The version information can be used as compile time constant or retrieved from a getter function at runtime.
```

File: ci/fpm.toml
Comment (ivan-pi) [inline ci/fpm.toml:2]: Is this to be understood as a reference to the version file? Wouldn't this complicate the version retrieval in the pretty output from fpm @LKedward is working on in https://github.com/fortran-lang/fpm/pull/622?

File: API-doc-FORD-file.md
Comment (ivan-pi) [inline API-doc-FORD-file.md:14]:
```suggestion
       PROJECT_VERSION_MAJOR=0
       PROJECT_VERSION_MINOR=1
       PROJECT_VERSION_PATCH=0
```

File: Makefile.manual
Comment (ivan-pi) [inline Makefile.manual:11]: Is the `shell cut` expected to run on Windows machines too?

File: ci/fpm.toml
Comment (awvwgk) [reply ci/fpm.toml:2]: No, the fpm model will see no difference between a version number specified in the manifest and a version number read from a file, since the deserializer will normalize this information. Using a version file will ensure we keep the numbers in sync everywhere.

File: API-doc-FORD-file.md
Comment (ivan-pi) [reply API-doc-FORD-file.md:14]: Or is this overridden during the fypp processing?

Comment (awvwgk) [reply API-doc-FORD-file.md:14]: This is supposed to be a stub, similar like the `MAXRANK=3` argument above, which would would not allow to actually compile stdlib.

File: doc/specs/stdlib_version.md
Comment (ivan-pi) [inline doc/specs/stdlib_version.md:21]:
```suggestion
String constant representing the version number.
```

File: Makefile.manual
Comment (awvwgk) [reply Makefile.manual:11]: Can the manual makefile run on Windows? I always thought this is a Unix-only thing.

**Review by ivan-pi — APPROVED** [review]
The CMake syntax to read a list is just terrible.

Pending the small issues, LGTM. Thanks.

File: Makefile.manual
Comment (ivan-pi) [reply Makefile.manual:11]: The Quickstart Fortran installer comes with Make included, but I haven't tested it. I remember your advice was ninja + CMake is the better approach for Windows. Personally I'm fine leaving as it is until things break.

File: ci/fpm.toml
Comment (ivan-pi) [reply ci/fpm.toml:2]: Okay, I've found it's mentioned here: https://github.com/fortran-lang/fpm/blob/main/manifest-reference.md#project-version

File: Makefile.manual
Comment (awvwgk) [reply Makefile.manual:11]: Windows is particular in this case, the `shell` command might not work at all if no POSIX shell is available on the platform, but it might pickup the git-bash which is also included in the installer and could contain other small Unix command line utilities.

I can have a look into parsing files in pure make, which should be possible, but the result might look even more convoluted than the CMake version string splitting ;)

Comment (brocolis) [reply Makefile.manual:11]: `shell cut` fails for me on Windows cmd.exe and Msys2, even with `cut.exe` in PATH

```
fypp -DMAXRANK=4 stdlib_optval.fypp stdlib_optval.f90
common.fypp:4: error: exception occured when setting variable(s) 'PROJECT_VERSION' to '"{}.{}.{}".format(PROJECT_VERSION_MAJOR, PROJECT_VERSION_MINOR, PROJECT_VERSION_PATCH)' [FyppFatalError]
error: name 'PROJECT_VERSION_MAJOR' is not defined [NameError]
gnumake[1]: *** [Makefile.manual:74: stdlib_optval.f90] Error 1
```

File: doc/specs/stdlib_version.md
Comment (brocolis) [inline doc/specs/stdlib_version.md:63]:
```suggestion
end program demo_version
```

**Review by brocolis — COMMENTED** [review]
@awvwgk many thanks for this PR. I tested it on Windows using CMake+Ninja.
Regarding the issue with $(shell cut ..), I'm trying to find a solution but no luck yet.

**Comment by brocolis** [comment]
OK I found a combination that works for me on Windows

``` 
set FC=gfortran
set FYPPFLAGS=-DMAXRANK=4
gnumake -f Makefile.manual
```

Thanks @awvwgk

**Review by brocolis — APPROVED** [review]
Thanks @awvwgk

**Comment by brocolis** [comment]
I just noticed `ubuntu-latest` CI is failing at `make -f Makefile.manual FYPPFLAGS="-DMAXRANK=4" -j`. It is the same issue. If we remove `FYPPFLAGS="-DMAXRANK=4"` from the command line and set it as environment variable, it will work, I think. That's how I did on my Windows.

File: Makefile.manual
Comment (awvwgk) [reply Makefile.manual]: Found a solution using the make commands `file`, `subst` and `word`.

File: API-doc-FORD-file.md
Comment (ivan-pi) [reply API-doc-FORD-file.md:14]: Don't these macros get applied when FORD runs fypp to preprocess the sources for the documentation? At least that's what I understand from here: https://github.com/Fortran-FOSS-Programmers/ford/wiki/Project-File-Options

File: Makefile.manual
Comment (ivan-pi) [inline Makefile.manual:7]:
```suggestion
VERSION := $(subst ., ,$(file < VERSION))
```
This is an input redirection right?

Comment (awvwgk) [reply Makefile.manual:7]: There is no space used in the make docs: https://www.gnu.org/software/make/manual/make.html#File-Function

File: API-doc-FORD-file.md
Comment (awvwgk) [reply API-doc-FORD-file.md:14]: It does show up in the final docs for the compile time constant:

image

Otherwise the value set won't affect the output of the documentation. I would prefer to leave this as a stub, so we don't have to worry about it becoming outdated when we release a new version.

File: doc/specs/stdlib_version.md
Comment (ivan-pi) [inline doc/specs/stdlib_version.md:52]: Just wondering if a fixed-length string can also be used to collect the output?

**Review by ivan-pi — APPROVED** [review]

File: Makefile.manual
Comment (ivan-pi) [reply Makefile.manual:7]: There seems to be a space here in the syntax definition `$(file op filename[,text])`, where op is one of >, <, <<. This is a nit-pick.

File: doc/specs/stdlib_version.md
Comment (awvwgk) [reply doc/specs/stdlib_version.md:52]: The dummy argument is currently `character(len=:), allocatable, intent(out), optional`, since we are using a subroutine rather than a function. The rationale is that we want a stable API *and* ABI for this procedure, we could also go for multiple functions for getting the individual values, to make use of it in a more functional way.

File: API-doc-FORD-file.md
Comment (ivan-pi) [reply API-doc-FORD-file.md:14]: Comments are not allowed in the FORD meta-data section. Is there somewhere we could document these values are stubs, and don't reflect the actual library?

**Review by brocolis — APPROVED** [review]
@awvwgk looks good.

**Comment by ivan-pi** [comment]
@jvdp1, @milancurcic, can one of you give this a third approval before merging?

File: doc/specs/stdlib_version.md
Comment (jvdp1) [inline doc/specs/stdlib_version.md:12]: Not sure about this, just a feeling:
```suggestion
The version information can be used as a compile time constant or retrieved from a getter function at runtime.
```

File: doc/specs/stdlib_version.md
Comment (jvdp1) [inline doc/specs/stdlib_version.md:14]: Why not only allow the correct approach to do it?

**Review by jvdp1 — APPROVED** [review]
LGTM. Just a couple of minor comments.
Thank you!

File: doc/specs/stdlib_version.md
Comment (awvwgk) [reply doc/specs/stdlib_version.md:14]: For the compact version I can see an application to verify the compatibility of API and/or ABI at runtime by a construct like:
```f90
use stdlib_version, only : stdlib_version_compact, get_stdlib_version
implicit none
integer :: major
call get_stdlib_version(major=major)
if (major > stdlib_version_compact / 10000) then
  error stop "Incompatible version of stdlib used"
end if
end
```

Similarly, `stdlib_version_string` and `get_stdlib_version(string=string)` can be used to provide a printout to inform about the stdlib versions used at build time and run time.

Not sure whether this is a realistic application of this information.

Comment (jvdp1) [reply doc/specs/stdlib_version.md:14]: Ok. Thank you for the explanation. You can merge this PR from my point-of-view.

**Review by milancurcic — APPROVED** [review]
Very nice addition, thank you.

Would it be useful for this module to also provide compile-time constants of compiler version and options used to build it?

```
module m
  use iso_fortran_env, only: compiler_version, compiler_options
  character(*), parameter :: stdlib_compiler_version = compiler_version()
  character(*), parameter :: stdlib_compiler_options = compiler_options()
end module m

program p
  use m
  print *, stdlib_compiler_version
  print *, stdlib_compiler_options
end
```

It works nicely with GFortran, but ifort 2021.3.0 doesn't seem to allow these intrinsic in constant expressions:

```
$ ifort test.f90 
test.f90(3): error #6263: This intrinsic function is invalid in constant expressions.   [COMPILER_VERSION]
  character(*), parameter :: stdlib_compiler_version = compiler_version()
-------------------------------------------------------^
test.f90(4): error #6263: This intrinsic function is invalid in constant expressions.   [COMPILER_OPTIONS]
  character(*), parameter :: stdlib_compiler_options = compiler_options()
-------------------------------------------------------^
test.f90(8): error #7013: This module file was not generated by any release of this compiler.   [M]
  use m
------^
compilation aborted for test.f90 (code 1)
```

which I think is incorrect. Anyhow, just an idea, something we could tackle in another PR.

**Comment by awvwgk** [comment]
Good idea, I opened a new issue for this one.

**Pull request merged**
